### DISCUSSION TITLE
KDL generator for fuzzing

### DISCUSSION THREAD

**Discussion by hkolbeck** [kdl-org/kdl#314] [General]
I wrote my first rust, a random valid KDL document generator for use in testing out parser implementations. It's very alpha-quality, but any reviews or feedback is quite welcome.

https://github.com/hkolbeck/kdl-gen

I think the first casualty is the spec itself. This rule in particular:

```
commented-block := '*/' | (multi-line-comment | '*' | '/' | [^*/]+) commented-block
```

The rule can resolve to `*` then `/` (or vice-versa), resulting in unbalanced multi-line comment boundaries.

There's probably a cleaner way, but this is the first thing that jumps into my head:

```
commented-block := '*/' | (multi-line-comment | '*[^/]' | '/[^*]' | [^*/]+) commented-block
```

If I'm missing something and this isn't a problem, let me know, otherwise I'll make an issue.

**Comment by Patitotective** [comment]
Looks nice, though, i think the best way to use it would be through GitHub actions, so rust and your library are installed automatically.
Looking forward to use it for [https://github.com/Patitotective/kdl-nim](kdl-nim) ;]

**Comment by zkat** [comment]
I mean, won't it hit that first `'*/'` case _before_ it hits a coincidental `*` followed by a `/`? I'm wondering if I'm thinking about it wrong, too, but I feel like this should be fine if the parser is going in the order specified.

**Reply by hkolbeck** [reply]
It's verging on a non-issue for sure. I don't think any parser will accept a document with mismatched boundaries, the issue is that the spec would technically say that such a document is legal KDL.

**Reply by zkat** [reply]
what I'm saying is that depending on how you interpret the grammar, it is, in fact, completely fine: The grammar was written with the assumption that clauses are checked *in order*, and with that assumption in place, you can't actually get unbalanced multiline comments.

**Comment by hkolbeck** [comment]
This has now found about 5 edge case bugs in https://github.com/hkolbeck/kdl4j. I've added options that I think will be useful for debugging, and I think this is more or less ready for usage by other parser authors.
